File: 01_数组/array/Demo_6.java
package array;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

/**
 * @author ZhaoWeiLong
 * @date 2021/3/12
 */
public class Demo_6 {

  public static void main(String[] args) {
    Integer[] arr = {4, 3, 2, 1};
    Integer[] arr2 = {9, 9, 9, 9};
    System.out.println(Arrays.toString(plusOne(arr)));
    System.out.println(Arrays.toString(plusOne(arr2)));
  }

  /**
   * 加一
   *
   * @param nums
   * @return
   */
  public static Integer[] plusOne(Integer[] nums) {

    ArrayList<Integer> list = new ArrayList<>();
    int addon = 0;

    for (int i = nums.length - 1; i >= 0; i--) {
      nums[i] += addon;
      addon = 0;
      if (i == nums.length - 1) {
        nums[i]++;
      }
      if (nums[i] == 10) {
        addon = 1;
        nums[i] = 0;
      }
    }
    if (addon == 1) {
      list.add(1);
      List<Integer> collect = Arrays.stream(nums).collect(Collectors.toList());
      list.addAll(collect);
      return list.toArray(new Integer[0]);
    } else {
      return nums;
    }
  }
}

File: 01_数组/array/Demo_6.go
package main

import "fmt"

func main() {
	//加一
	/**
	最高位数字存放在数组的首位， 数组中每个元素只存储单个数字。你可以假设除了整数 0 之外，这个整数不会以零开头。

	示例 1:

	输入: [1,2,3]
	输出: [1,2,4]
	解释: 输入数组表示数字 123。
	示例 2:

	输入: [4,3,2,1]
	输出: [4,3,2,2]
	解释: 输入数组表示数字 4321。
	*/
	nums := []int{4, 3, 2, 1}
	nums2 := []int{9, 9, 9, 9}
	fmt.Println(plusOne(nums))
	fmt.Println(plusOne(nums2))

}

func plusOne(nums []int) []int {
	var result []int
	addon := 0
	for i := len(nums) - 1; i >= 0; i-- {
		//在进位的基础上再加一
		nums[i] += addon
		//重置进位
		addon = 0
		//如果是第一次，直接加一
		if i == len(nums)-1 {
			nums[i]++
		}
		//如果加以等于10，则需要进位处理
		if nums[i] == 10 {
			addon = 1
			nums[i] = 0
		}
	}
	if addon == 1 {
		result = make([]int, 1)
		result[0] = 1
		result = append(result, nums...)
	} else {
		result = nums
	}
	return result
}
